File: domain-persistence-modeling/java-spring-boot/src/main/java/x/scratch/child/ChildFactory.java
package x.scratch.child;

import java.util.Optional;
import java.util.stream.Stream;

public interface ChildFactory {
    Stream<Child> all();

    Optional<Child> findExisting(String naturalId);

    Child createNewUnassigned(String naturalId);

    Child findExistingOrCreateNewUnassigned(String naturalId);

    Stream<Child> findOwned(String parentNaturalId);
}

File: spring-boot-domain-persistence-modeling-kotlin/src/main/kotlin/x/domainpersistencemodeling/java/ChildFactory.kt
package x.domainpersistencemodeling.java

import java.util.stream.Stream

interface ChildFactory {
    fun all(): Stream<Child>
    fun findExisting(naturalId: String): Child?
    fun createNew(naturalId: String): Child
    fun findExistingOrCreateNew(naturalId: String): Child
    fun byParentNaturalId(parentNaturalId: String): Sequence<Child>
}
